proposals/NNNN-method-keypaths.md

```diff
@@ -0,0 +1,643 @@
+  proposal and not in conflict with it.
+- If we seek a broader solution, it would need to be consistent with unbound methods as well, and thus (as in the
+  previous section) this proposal is only uniformly respecting an existing problem, not introducing a new one.
+
+### Support for argument labels
+
+Accepting this proposal will create pressure to solve the argument label problem in the
+[wrapper type example above](#abstractions-over-properties-and-methods).
+
+One possibility would be loosening SE-0111 to make argument labels optional:
+
+```swift
+struct Foo {
+    func bar(a: A) { ... }
+ }
+
+struct FooWrapper { ... }
+let wrappedFoo: FooWrapper = ...
+
+wrappedFoo.bar(a: A())
+// this would be equivalent to:
+wrappedFoo.bar(a:)(A())
+```
+
+Alternatively, a
+[recent pitch](https://forums.swift.org/t/include-argument-labels-in-identifiers/35367/23) suggested that argument
+labels be added to the identifier instead of the function type. It might be possible to extend this to identifiers
+dynamically resolved by key paths.
+
+However, given that SE-0111 and its implications have generated copious discussion without resolution
+([1](https://forums.swift.org/t/review-se-0111-remove-type-system-significance-of-function-argument-labels/3209),
+[2](https://forums.swift.org/t/update-commentary-se-0111-remove-type-system-significance-of-function-argument-labels/3391),
+[3](https://forums.swift.org/t/accepted-se-0111-remove-type-system-significance-of-function-argument-labels/3306/17),
+[4](https://forums.swift.org/t/se-0111-related-question/3813),
+[5](https://forums.swift.org/t/se-0111-and-curried-argument-labels-unintended-consequences/4179)),
+trying to solve the argument label problem as part of this proposal would likely doom it — and its benefits — to
+purgatory.
+
+
+## Alternatives considered
+
+### Focus on unbound methods instead of key paths
+
+This proposal extends the domain of key paths. We could instead extend the domain of unbound methods:
+
+```swift
+String.count  // yields a (String) -> Int function
+```
+
+This however does not provide the [contextual type inference benefits of key paths](#idioms-that-rely-on-contextual-type),
+departs from `@dynamicMemberLookup`’s current focus on key paths, and would create confusing error messages for
+developers who confuse instance properties with type properties.
+
+### Prohibit method key paths in dynamic member lookup
+
+This would prevent wrappers from exposing label-stripped methods:
+
+```swift
+let wrapper: SomeStringWrapper = ...
+
+wrapper.hash(&value) // ❌ Don't allow this until language supports `into:` label
+
+wrapper.dropFirst()  // ❌ Also wouldn’t be allowed
+```
+
+The argument in favor of this is it would be best not to expose methods stripped of labels now if we expect the labels
+to return in the future, since their re-addition _might_ be source-breaking.
```

Comment:
I think that focusing on coherence and consistency between the wrapper and the original type is better so as not to confuse the reader with source breaking changes. I do like, however, you mentioning that this problem already exists for unbound methods in general and as such solutions would likely be source-breaking. Additionally you should mention that we chose not to prohibit the use of closures through dynamic member lookup, because this is intended to be a small and additive proposal. Instead, what the alternative would require is that this proposal be source-breaking.

Reply:
You make a good point that it’s important to show `hash(into:)` here to make the inconsistency clear.

>Additionally you should mention that we chose not to prohibit the use of closures through dynamic member lookup, because this is intended to be a small and additive proposal. Instead, what the alternative would require is that this proposal be source-breaking.

Why would the alternative be source-breaking? Key path dynamic member lookup doesn’t support methods now, so prohibiting them in the proposal wouldn’t prohibit anything that works today. I think I missed something?